lc-hax/Scripts/Modules/Possession/PossessionMod.cs

```diff
@@ -190,15 +240,146 @@ internal sealed class PossessionMod : MonoBehaviour {
+        if (this.EnemyControllers.TryGetValue(enemy.GetType(), out IController value))
+            this.HandleEnemyOnUnpossess(value, enemy);
+
+        this.IsAiControlled = false;
+        this.ResetInteractionCooldowns();
         this.Possession.Clear();
     }
 
+    internal void ToggleAiControl() {
+        if (this.Possession.Enemy is null || this.Possession.Enemy.agent.Unfake() is null ||
+            this.CharacterMovement is null || this.MousePan is null) return;
+        this.IsAiControlled = !this.IsAiControlled;
+        this.SetAiControl(this.IsAiControlled, true);
+    }
+
+    internal void SetAiControl(bool EnableAI, bool DisplayNotification = false) {
+        if (this.Possession.Enemy is not EnemyAI enemy) return;
+        if (enemy.agent.Unfake() is not NavMeshAgent navMeshAgent) return;
+        if (this.CharacterMovement is not CharacterMovement characterMovement) return;
+
+        if (EnableAI) {
+            _ = enemy.agent.Warp(enemy.transform.position);
+            enemy.SyncPositionToClients();
+        }
+
+        if (this.NoClipEnabled) {
+            this.NoClipEnabled = false;
+            characterMovement.SetNoClipMode(false);
+        }
+
+        navMeshAgent.updatePosition = EnableAI;
+        navMeshAgent.updateRotation = EnableAI;
+        navMeshAgent.isStopped = !EnableAI;
+        characterMovement.SetPosition(enemy.transform.position);
+        characterMovement.enabled = !EnableAI;
+        if (DisplayNotification) {
+            Helper.SendNotification(
+                title: "AI Control:",
+                body: this.IsAiControlled ? "Enabled" : "Disabled"
+            );
+        }
+
+    }
+
+    public void ResetInteractionCooldowns() {
+        this.teleportCooldownRemaining = -DoorInteractionCooldown;
+        this.doorCooldownRemaining = -TeleportDoorCooldown;
+    }
+
+
+    bool CanUseEntranceDoors(EnemyAI enemy) {
+        return enemy is not EnemyAI enemyAI
+            ? false
+            : this.EnemyControllers.TryGetValue(enemy.GetType(), out IController value) && value.CanUseEntranceDoors(enemyAI);
+    }
+
+    bool SyncAnimationSpeedEnabled(EnemyAI enemy) {
+        return enemy is not EnemyAI enemyAI
+            ? false
+            : this.EnemyControllers.TryGetValue(enemy.GetType(), out IController value) && value.SyncAnimationSpeedEnabled(enemyAI);
+    }
+
+
+    float InteractRange(EnemyAI enemy) {
+        return enemy is not EnemyAI enemyAI
+            ? 0
+            : this.EnemyControllers.TryGetValue(enemy.GetType(), out IController value)
+            ? value.InteractRange(enemyAI).GetValueOrDefault(2.5f)
+            : 2.5f;
+    }
+
+    float SprintMultiplier(EnemyAI enemy) {
+        return enemy is not EnemyAI enemyAI
+            ? 0
+            : this.EnemyControllers.TryGetValue(enemy.GetType(), out IController value)
+            ? value.SprintMultiplier(enemyAI).GetValueOrDefault(2.8f)
+            : 2.8f;
+    }
```

Comment:
This is so cursed. You don't even need to do it. It's why we have default implementations in our interface.

Reply:
Nevermind, I am wrong here. I am not sure how I can do this better.